crates/ruff/src/noqa.rs

```diff
@@ -255,41 +273,202 @@ enum ParsedFileExemption<'a> {
 
 impl<'a> ParsedFileExemption<'a> {
     /// Return a [`ParsedFileExemption`] for a given comment line.
-    fn try_extract(line: &'a str) -> Option<Self> {
-        let line = line.trim_whitespace_start();
+    fn try_extract(line: &'a str) -> Result<Option<Self>, ParseError> {
+        let line = ParsedFileExemption::lex_whitespace(line);
```

Comment:
What could help to clean this up potentially is to implement `ParsedFileExemption` as a "lexer" which returns different Tokens:

```
enum ExemptionToken {
	Whitespace,
	Flake8,
	Hash,
	Noqa,
	Code,
```

See `SimpleTokenizer` with its `Cursor` implementation for inspiration. `try_extract` is then a parser where you could have `eat` etc functions. But this may be very overkill.

Reply:
Yeah in hindsight perhaps I should've done this. I may look into it as a follow-up PR.